File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int i = 1;
        ListNode* slow = head, *fast = head;
        ListNode* prev = NULL;
        for( ; i < n; fast = fast->next, i++);
        while(fast && fast->next){
            prev = slow;
            slow = slow->next;
            fast = fast->next;
        }
        if(prev) prev->next = slow->next;
        else head = slow->next;
        return head;
    }
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        int count = 0;
        ListNode temp = head;
        if(head == null || head.next == null) return null;
        while(temp != null){
            count++;
            temp = temp.next;
        }
        if(count == n){
            head = head.next;
            return head;
        }
        int del = count - n;
        int pos = 1;
        temp = head;
        while(pos < del){
            temp = temp.next;
            pos++;
        }
        temp.next = temp.next.next;
        return head;
    }
}